src/Compilers/CSharp/Test/Syntax/Parsing/ForStatementParsingTest.cs

```diff
@@ -448,4 +443,2690 @@ public void TestVariableDeclaratorVersusCondition1()
+            {
+                N(SyntaxKind.DotDotToken);
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+            }
+            N(SyntaxKind.CloseParenToken);
+            N(SyntaxKind.EmptyStatement);
+            {
+                N(SyntaxKind.SemicolonToken);
+            }
+        }
+        EOF();
+    }
+
+    [Fact]
+    public void TestVariousExpressions_Range4()
+    {
+        UsingStatement("""
+            for (a..a;a..a;a..a);
+            """);
+
+        N(SyntaxKind.ForStatement);
+        {
+            N(SyntaxKind.ForKeyword);
+            N(SyntaxKind.OpenParenToken);
+            N(SyntaxKind.RangeExpression);
+            {
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+                N(SyntaxKind.DotDotToken);
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+            }
+            N(SyntaxKind.SemicolonToken);
+            N(SyntaxKind.RangeExpression);
+            {
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+                N(SyntaxKind.DotDotToken);
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+            }
+            N(SyntaxKind.SemicolonToken);
+            N(SyntaxKind.RangeExpression);
+            {
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+                N(SyntaxKind.DotDotToken);
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+            }
+            N(SyntaxKind.CloseParenToken);
+            N(SyntaxKind.EmptyStatement);
+            {
+                N(SyntaxKind.SemicolonToken);
+            }
+        }
+        EOF();
+    }
+
+    [Fact]
+    public void TestVariousExpressions_Ref1()
+    {
+        UsingStatement("""
+            for (ref a; ref a; ref a);
+            """,
+            // (1,11): error CS1001: Identifier expected
+            // for (ref a; ref a; ref a);
+            Diagnostic(ErrorCode.ERR_IdentifierExpected, ";").WithLocation(1, 11),
+            // (1,13): error CS1525: Invalid expression term 'ref'
+            // for (ref a; ref a; ref a);
+            Diagnostic(ErrorCode.ERR_InvalidExprTerm, "ref a").WithArguments("ref").WithLocation(1, 13),
+            // (1,20): error CS1525: Invalid expression term 'ref'
+            // for (ref a; ref a; ref a);
+            Diagnostic(ErrorCode.ERR_InvalidExprTerm, "ref a").WithArguments("ref").WithLocation(1, 20));
```

Comment:
also surprising, but not really a problem.  the strict ang grammar truly does not allow 'ref' here.  We could support parsing this, but would have to then ensure binding errors. a s this is extremely unlikely to be written, no need to do extra work here.